src/tools/run-make-support/src/assertion_helpers/mod.rs

```diff
@@ -0,0 +1,160 @@
+}
+
+struct SearchDetails<'assertion_name, 'haystack, 'query> {
+    assertion_name: &'assertion_name str,
+    haystack: &'haystack str,
+    query_kind: &'static str,
+    query: &'query str,
+}
+
+impl<'assertion_name, 'haystack, 'query> SearchDetails<'assertion_name, 'haystack, 'query> {
+    fn dump(&self) {
+        eprintln!("{}:", self.assertion_name);
+        eprintln!("=== HAYSTACK ===");
+        eprintln!("{}", self.haystack);
+        eprintln!("=== {} ===", self.query_kind);
+        eprintln!("{}", self.query);
+    }
+}
+
+/// Assert that `haystack` contains `needle`.
+#[track_caller]
+pub fn assert_contains<H: AsRef<str>, N: AsRef<str>>(haystack: H, needle: N) {
+    let haystack = haystack.as_ref();
+    let needle = needle.as_ref();
+    if !haystack.contains(needle) {
+        SearchDetails {
+            assertion_name: "assert_contains",
+            haystack,
+            query_kind: "NEEDLE",
+            query: needle,
+        }
+        .dump();
+        panic!("needle was not found in haystack");
+    }
+}
+
+/// Assert that `haystack` does not contain `needle`.
+#[track_caller]
+pub fn assert_not_contains<H: AsRef<str>, N: AsRef<str>>(haystack: H, needle: N) {
+    let haystack = haystack.as_ref();
+    let needle = needle.as_ref();
+    if haystack.contains(needle) {
+        SearchDetails {
+            assertion_name: "assert_not_contains",
+            haystack,
+            query_kind: "NEEDLE",
+            query: needle,
+        }
+        .dump();
+        panic!("needle was unexpectedly found in haystack");
+    }
+}
+
+/// Assert that `haystack` contains the regex `pattern`.
+#[track_caller]
+pub fn assert_contains_regex<H: AsRef<str>, N: AsRef<str>>(haystack: H, pattern: N) {
+    let haystack = haystack.as_ref();
+    let pattern = pattern.as_ref();
+    let re = regex::Regex::new(pattern).unwrap();
+    if !re.is_match(haystack) {
+        SearchDetails {
+            assertion_name: "assert_contains_regex",
+            haystack,
+            query_kind: "REGEX",
+            query: pattern,
+        }
+        .dump();
+        panic!("regex was not found in haystack");
+    }
+}
+
+/// Assert that `haystack` does not contain the regex `pattern`.
+#[track_caller]
+pub fn assert_not_contains_regex<H: AsRef<str>, N: AsRef<str>>(haystack: H, pattern: N) {
+    let haystack = haystack.as_ref();
+    let pattern = pattern.as_ref();
+    let re = regex::Regex::new(pattern).unwrap();
+    if re.is_match(haystack) {
+        SearchDetails {
+            assertion_name: "assert_not_contains_regex",
+            haystack,
+            query_kind: "REGEX",
+            query: pattern,
+        }
+        .dump();
+        panic!("regex was unexpectedly found in haystack");
+    }
+}
+
+/// Assert that `haystack` contains regex `pattern` an `expected_count` number of times.
+#[track_caller]
+pub fn assert_count_is<H: AsRef<str>, N: AsRef<str>>(
+    expected_count: usize,
+    haystack: H,
+    pattern: N,
```

Comment:
Renaming `pattern` to `needle` but only for some of these functions is a bit inconsistent 🤔

Reply:
This was intended to distinguish between the assert helpers that accept a plain-text substring vs regex patterns.

Reply:
Reverted these renames, just used `needle` consistently.